File: tests/pages/login.page.ts
import { Page, Locator, expect } from '@playwright/test';
import data from '../data/data.json';

export class LoginPage {

  constructor(private page: Page) {}

  async open() {
    await this.page.goto(data.loginPage);
  }
  
  usernameInput() { return this.page.locator('input[data-test="username"]'); }
  passwordInput() { return this.page.locator('input[data-test="password"]'); }
  loginButton()   { return this.page.locator('input[data-test="login-button"]'); }
  errorMessage()  { return this.page.locator('[data-test="error"]'); }

  async login(username: string, password: string) {
    await this.usernameInput().fill(username);
    await this.passwordInput().fill(password);
    await this.loginButton().click();
  }

  async expectErrorVisible(error: string) {
    await expect(this.errorMessage()).toContainText(error);
  }

  async expectOnInventoryPage() {
    await expect(this.page).toHaveURL(/inventory\.html/);
  }
}

File: tests/pages/login.page.cs
// LoginPage.cs
using System.Threading.Tasks;
using Microsoft.Playwright;
using static Microsoft.Playwright.Assertions;

namespace MyTests.Pages;

public sealed class LoginPage
{
    private readonly IPage _page;
    private readonly string _loginPageUrl;

    public LoginPage(IPage page, string loginPageUrl)
    {
        _page = page;
        _loginPageUrl = loginPageUrl;
    }

    // NAVIGATION
    public Task OpenAsync() => _page.GotoAsync(_loginPageUrl);

    // LOCATORS
    public ILocator UsernameInput => _page.Locator("input[data-test=\"username\"]");
    public ILocator PasswordInput => _page.Locator("input[data-test=\"password\"]");
    public ILocator LoginButton   => _page.Locator("input[data-test=\"login-button\"]");
    public ILocator ErrorMessage  => _page.Locator("[data-test=\"error\"]");

    // ACTIONS
    public async Task LoginAsync(string username, string password)
    {
        await UsernameInput.FillAsync(username);
        await PasswordInput.FillAsync(password);
        await LoginButton.ClickAsync();
    }

    // VALIDATIONS
    public async Task ExpectErrorVisibleAsync(string errorText)
    {
        await Expect(ErrorMessage).ToContainTextAsync(errorText);
    }

    public async Task ExpectOnInventoryPageAsync()
    {
        await Expect(_page).ToHaveURLAsync(new System.Text.RegularExpressions.Regex("inventory\\.html"));
    }
}
